adding a "block" func to the template's FuncMap

arschles opened the pull request:
allows multiple `yield` statements in a layout.

a sample `layout.tmpl`:

```html
<html>
<head>
{{yield "title"}}
{{yield "css"}}
</head>
<body>
{{yield "body"}}
</body>
</html>
```

a sample `page.html`:

```html
{{define "title-page"}}<title>hello world!</title>{{end}}
{{define "css-page"}}<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.5/css/bootstrap.min.css"/>{{end}}
{{define "js-page}}<script type="text/javascript" src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.5/js/bootstrap.min.js"></script>{{end}}
```

added tests but many others are failing also

unrolled commented:
Hmm, I see what you're trying to do, but this would break the current API implementation (as seen in the failed tests). I'll think about an alternative approach, but feel free to suggestion something in the meantime.

arschles commented:
yep. one option (and I'm not sure that this even qualifies as a valid option) is to add a `block` function to the `FuncMap`. thoughts?

panjunjie commented:
+1 Good idea.

arschles commented:
@unrolled @panjunjie added the `block` function + tests & docs for it, seems to work ok. Also renamed `addYield` to `addLayoutFuncs` to indicate its more general purpose.

Finally, I changed the `r.templates.Funcs(...)` call to do a template lookup first and call `Funcs` on that specific template. I did this because `Funcs` overwrites the map which I think can cause a race with `name` in the `HTML` func. That is unless I'm missing some invariant somewhere. Looking for feedback there.

unrolled commented on README.md:
```diff
@@ -174,10 +174,16 @@ r := render.New(render.Options{
 <html>
   <head>
     <title>My Layout</title>
+    <!-- Render the template called `css-$current_template` here -->
```
`<!-- Render the block template called `css-$current_template` here -->`

unrolled commented on README.md:
```diff
@@ -174,10 +174,16 @@ r := render.New(render.Options{
+    {{ block "css" }}
   </head>
   <body>
+    <!-- render the template called `header-$current_template` here -->
```
`<!-- Render the block template called 'header-$current_template' here -->`

unrolled commented on README.md:
```diff
@@ -174,10 +174,16 @@ r := render.New(render.Options{
+    {{ block "header" }}
     <!-- Render the current template here -->
     {{ yield }}
+    <!-- render the template called `footer-$current_template` here -->
```
`<!-- Render the block template called 'footer-$current_template' here -->`

unrolled commented on README.md:
```diff
@@ -84,7 +84,7 @@ r := render.New(render.Options{
       return []string{"filename.tmpl"}
     },
-    Layout: "layout", // Specify a layout template. Layouts can call {{ yield }} to render the current template.
+    Layout: "layout", // Specify a layout template. Layouts can call {{ yield "blockname" }} to render a block in the current template.
```
This needs to change again as well. Something like:
`// Specify a layout template. Layouts can call {{ yield }} to render a block in the current template, or call {{ block "css" }} to render a template block.`

unrolled commented:
Couple of tweaks, but looks good!

arschles commented:
@unrolled fixed the tweaks. thanks!
